### DISCUSSION TITLE
String from print() comes as Unicode while from raise Exception() as AnsiString

### DISCUSSION THREAD

**Discussion by MKostitsyn** [pyscripter/python4delphi#465] [Q&A] [answered]
I have Unicode file sample.py with the following script inside:
```
print("Print message: АБВГД ĀČĶ āčķ")
raise Exception("Raising exception: АБВГД ĀČĶ āčķ")
```
In IO UnicodeIO is set to True.

In OnSendUnicodeData() I simply dump all received messages into a text file. That's how it looks when I run the script:

![image](https://github.com/pyscripter/python4delphi/assets/9676383/ebc7d55b-ef68-4b01-bac7-542b606ff110)

Some of the data comes in incorrect encoding.

Any ideas how to solve this?

**Comment by MKostitsyn** [comment]
In python IDLE it works fine:
![image](https://github.com/pyscripter/python4delphi/assets/9676383/98498b28-ba9d-4860-b98c-9268455f9e07)

**Comment by pyscripter** [comment]
When I run your code with Demo01 I works well.

![image](https://github.com/pyscripter/python4delphi/assets/1311616/4836fd43-76e1-42da-b5ca-c14ced859fb4)


I used 
```pythoon
PythonEngine1.ExecStrings( Memo1.Lines, 'c:\temp\sample.py' );
```
**sample.py is utf8 encoded.**

Could you post a simple project in a zip file that reproduces the issue?

**Comment by MKostitsyn** [comment]
[PythonSample.zip](https://github.com/pyscripter/python4delphi/files/14788051/PythonSample.zip)

Here.

**Comment by MKostitsyn** [comment]
![image](https://github.com/pyscripter/python4delphi/assets/9676383/8cc53384-3c5b-4788-91f2-10dd01455490)

The script should be in the file and the script file name should be passed to the ExecStrings method as a parameter to reproduce the mistake.

**Comment by pyscripter** [comment, marked as answer]
I have not tested but you need to specify a UTF8 encoding in 

Script.LoadFromFile(aFileName);

and

TInputOutputStub.Output.SaveToFile('c:\Devel\tmp\PythonSample\sample.out')

**Reply by MKostitsyn** [reply]
When I specify the encoding in LoadFromFile, the following lines disappear from the output (which is really weird), and hence there is no encoding problem, not because the problem is solved, but because there is no line where the problem appeared previously.
I would like to have full information about the exceptions I receive in my scripts.

```
    raise Exception("Raising exception: АБВГД ĀČĶ āčķ")
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Reply by pyscripter** [reply]
Here it works fine with:

      Script.LoadFromFile(aFileName, TEncoding.UTF8);

      TInputOutputStub.Output.SaveToFile('c:\users\username\downloads\PythonSample\sample.out', TEncoding.UTF8);


sample,out

```
Print message: АБВГД ĀČĶ āčķ
Traceback (most recent call last):
  File "c:\users\kiria\downloads\PythonSample\sample.py", line 2, in <module>
    raise Exception("Raising exception: АБВГД ĀČĶ āčķ")
Exception: Raising exception: АБВГД ĀČĶ āčķ
Unhandled exception caught: EPyException: Exception: Raising exception: АБВГД ĀČĶ āčķ
```

**Reply by MKostitsyn** [reply]
![image](https://github.com/pyscripter/python4delphi/assets/9676383/c23428b9-debc-4c1d-8d48-d42ef2966c4f)
Why did this line disappear?

**Reply by pyscripter** [reply]
This line does not come up in either IDLE or when you run the script with python
python sample.py

The output I am getting is identical to that of IDLE and python.

**Reply by MKostitsyn** [reply]
Yes, but if you remove encoding, this line appears and it shows the exact cause of the exception.
For my project this information would be useful.

**Comment by pyscripter** [comment]
> if you remove encoding

the code you run is not the code you indent to run.  The code you indent to run produces the output without that line.

**Reply by MKostitsyn** [reply]
Did you try running the sample project without specifying the encoding?

**Reply by pyscripter** [reply]
Yes.   And if you inspect Script.Text you will see it is wrong.

**Comment by pyscripter** [comment]
You get that line in other cases e,g,

sample.py
print "Hi"

sample.out
```
  File "c:\users\kiria\downloads\PythonSample\sample.py", line 1
    print "Hi"
    ^^^^^^^^^^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)?
Unhandled exception caught: EPySyntaxError: SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)? (line 1, offset 1): 'print "Hi"'
```

**Reply by MKostitsyn** [reply]
What if you add some national symbols to the line? 
For example:
`print "Hi АБВГД ĀČĶ āčķ"`

**Reply by pyscripter** [reply]
Why don't you try it?

**Reply by MKostitsyn** [reply]
As I understand we receive different results in some cases.

**Reply by pyscripter** [reply]
Not if you encode your files.   Python expects and uses utf8 encoding,  In any case this is not a P4D issue.

**Reply by pyscripter** [reply]
But since you asked:

sample.out
```
  File "c:\users\kiria\downloads\PythonSample\sample.py", line 1
    print "Hi АБВГД ĀČĶ āčķ"
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)?
Unhandled exception caught: EPySyntaxError: SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)? (line 1, offset 1): 'print "Hi АБВГД ĀČĶ āčķ"'

```

**Reply by MKostitsyn** [reply]
The behavioral is very strange.

I now specify the encoding as you suggested and run the script from zip above and I don't see the "^^^" line:
![image](https://github.com/pyscripter/python4delphi/assets/9676383/5e45f2c1-5d3c-411f-aa41-c1f496e54b80)

If I add space after the Exception(), then the "^^^" line appears:
![image](https://github.com/pyscripter/python4delphi/assets/9676383/6c23fffe-c01b-46e8-ae9d-8da754a71cba)

I believe it's not a P4D issue, thanks for your help!

**Comment by MKostitsyn** [comment]
The last thing.
Maybe you will want to specify encoding here in P4D library.

```
procedure TPythonEngine.ExecFile(const FileName: string; locals,
  globals: PPyObject);
var
  SL: TStringList;
begin
  SL := TStringList.Create;
  try
    SL.LoadFromFile(FileName);
    ExecStrings(SL, locals, globals, FileName);
  finally
    SL.Free;
  end;
end;
```

**Reply by pyscripter** [reply]
Done.
